### ISSUE TITLE
Windows installer doesn't properly overwrite itself

### ISSUE THREAD

**Issue by pwn0r** [certbot/certbot#8824]
## My operating system is (include version):
Windows 2012 R2

## I installed Certbot with (snap, OS package manager, pip, certbot-auto, etc):

beta download from website https://dl.eff.org/certbot-beta-installer-win32.exe

## Certbot's behavior differed from what I expected because:
cryptography problem, but depending on actual version can be different

I use certbot only to generate certs, I install them myself on the webservers

Had iirc 1.12 version installed, downloaded and installed 1.14 (? if thats' the current one) -- run into cryptography error. 
If you had a different version then a different problem can be reported (i run some tests with a few older versions I had).

Basically it seems that in my case /pkgs subfolder contained both versions of cryptography (2.8 and 3.4.5 iirc), which triggered an exception when running certbot. 

TLDR -- it seems installing a newer beta version by installer may cause a conflict in python installation. 

In my case there was a relatively easy workaround: I deleted pkgs subfolder manually and run the installer again. After that it worked as expected.

**Comment by alexzorin** [comment]
Thanks, confirmed.

<details>
<summary>Stack trace</summary>

    C:\Program Files (x86)\Certbot>certbot --version
    Traceback (most recent call last):
    File "runpy.py", line 194, in _run_module_as_main
    File "runpy.py", line 87, in _run_code
    File "C:\Program Files (x86)\Certbot\bin\certbot.exe\__main__.py", line 17, in <module>
    File "C:\Program Files (x86)\Certbot\pkgs\certbot\main.py", line 2, in <module>
    from certbot._internal import main as internal_main
    File "C:\Program Files (x86)\Certbot\pkgs\certbot\_internal\main.py", line 17, in <module>
    import josepy as jose
    File "C:\Program Files (x86)\Certbot\pkgs\josepy\__init__.py", line 41, in <module>
    from josepy.interfaces import JSONDeSerializable
    File "C:\Program Files (x86)\Certbot\pkgs\josepy\interfaces.py", line 6, in <module>
    from josepy import errors, util
    File "C:\Program Files (x86)\Certbot\pkgs\josepy\util.py", line 4, in <module>
    import OpenSSL
    File "C:\Program Files (x86)\Certbot\pkgs\OpenSSL\__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL
    File "C:\Program Files (x86)\Certbot\pkgs\OpenSSL\crypto.py", line 17, in <module>
    from OpenSSL._util import (
    File "C:\Program Files (x86)\Certbot\pkgs\OpenSSL\_util.py", line 6, in <module>
    from cryptography.hazmat.bindings.openssl.binding import Binding
    File "C:\Program Files (x86)\Certbot\pkgs\cryptography\hazmat\bindings\openssl\binding.py", line 170, in <module>
    _verify_package_version(cryptography.__version__)
    File "C:\Program Files (x86)\Certbot\pkgs\cryptography\hazmat\bindings\openssl\binding.py", line 159, in _verify_package_version
    raise ImportError(
    ImportError: The version of cryptography does not match the loaded shared object. This can happen if you have multiple copies of cryptography installed in your Python path. Please try creating a new virtual environment to resolve this issue. Loaded python version: 3.4.7, shared object version: b'2.8'
</details>

<details>
<summary>Resulting `pkgs` directory after upgrading 1.12 to 1.15:</summary>

C:\Program Files (x86)\Certbot\pkgs>dir
Volume in drive C has no label.
Volume Serial Number is 524F-0F04

Directory of C:\Program Files (x86)\Certbot\pkgs

05/06/2021 12:37 AM <DIR> .
05/06/2021 12:37 AM <DIR> ..
05/06/2021 12:33 AM <DIR> acme
05/06/2021 12:33 AM <DIR> acme-1.12.0.dist-info
05/06/2021 12:37 AM <DIR> acme-1.15.0.dist-info
05/06/2021 12:33 AM <DIR> adodbapi
05/06/2021 12:33 AM <DIR> certbot
05/06/2021 12:33 AM <DIR> certbot-1.12.0.dist-info
05/06/2021 12:37 AM <DIR> certbot-1.15.0.dist-info
05/06/2021 12:33 AM <DIR> certifi
05/06/2021 12:37 AM <DIR> certifi-2020.12.5.dist-info
05/06/2021 12:33 AM <DIR> certifi-2020.4.5.1.dist-info
05/06/2021 12:33 AM <DIR> cffi
05/06/2021 12:33 AM <DIR> cffi-1.14.0.dist-info
05/06/2021 12:37 AM <DIR> cffi-1.14.5.dist-info
05/06/2021 12:37 AM <DIR> chardet
05/06/2021 12:33 AM <DIR> chardet-3.0.4.dist-info
05/06/2021 12:37 AM <DIR> chardet-4.0.0.dist-info
05/06/2021 12:33 AM <DIR> ConfigArgParse-1.2.3.dist-info
05/06/2021 12:37 AM <DIR> ConfigArgParse-1.4.dist-info
05/05/2021 12:45 AM 47,242 configargparse.py
05/06/2021 12:33 AM <DIR> configobj-5.0.6.dist-info
05/05/2021 12:45 AM 89,627 configobj.py
05/06/2021 12:37 AM <DIR> cryptography
05/06/2021 12:33 AM <DIR> cryptography-2.8.dist-info
05/06/2021 12:37 AM <DIR> cryptography-3.4.7.dist-info
05/06/2021 12:33 AM <DIR> distro-1.5.0.dist-info
05/05/2021 12:45 AM 43,628 distro.py
05/05/2021 12:45 AM 152 distutils-precedence.pth
02/02/2021 07:10 PM 126 easy_install.py
05/06/2021 12:33 AM <DIR> idna
05/06/2021 12:37 AM <DIR> idna-2.10.dist-info
05/06/2021 12:33 AM <DIR> idna-2.9.dist-info
05/06/2021 12:33 AM <DIR> isapi
05/06/2021 12:33 AM <DIR> josepy
05/06/2021 12:33 AM <DIR> josepy-1.3.0.dist-info
05/06/2021 12:37 AM <DIR> josepy-1.8.0.dist-info
05/06/2021 12:33 AM <DIR> OpenSSL
05/06/2021 12:33 AM <DIR> parsedatetime
05/06/2021 12:33 AM <DIR> parsedatetime-2.5.dist-info
05/06/2021 12:37 AM <DIR> parsedatetime-2.6.dist-info
05/06/2021 12:37 AM <DIR> pkg_resources
05/05/2021 12:45 AM 64 placeholder
05/06/2021 12:33 AM <DIR> pycparser
05/06/2021 12:33 AM <DIR> pycparser-2.20.dist-info
05/06/2021 12:33 AM <DIR> pyOpenSSL-19.1.0.dist-info
05/06/2021 12:37 AM <DIR> pyOpenSSL-20.0.1.dist-info
05/06/2021 12:33 AM <DIR> pyrfc3339
05/06/2021 12:33 AM <DIR> pyRFC3339-1.1.dist-info
05/05/2021 12:45 AM 138 pythoncom.py
05/06/2021 12:33 AM <DIR> pythonwin
05/06/2021 12:33 AM <DIR> pytz
05/06/2021 12:33 AM <DIR> pytz-2020.1.dist-info
05/06/2021 12:37 AM <DIR> pytz-2021.1.dist-info
05/06/2021 12:33 AM <DIR> pywin32-300.dist-info
05/05/2021 12:45 AM 2,666,406 PyWin32.chm
05/05/2021 12:45 AM 178 pywin32.pth
05/05/2021 12:45 AM 5 pywin32.version.txt
05/06/2021 12:33 AM <DIR> pywin32_system32
05/06/2021 12:33 AM <DIR> requests
05/06/2021 12:33 AM <DIR> requests-2.23.0.dist-info
05/06/2021 12:37 AM <DIR> requests-2.25.1.dist-info
05/06/2021 12:33 AM <DIR> requests_toolbelt
05/06/2021 12:33 AM <DIR> requests_toolbelt-0.9.1.dist-info
05/06/2021 12:37 AM <DIR> setuptools
05/06/2021 12:33 AM <DIR> setuptools-44.1.1.dist-info
05/06/2021 12:37 AM <DIR> setuptools-54.1.2.dist-info
05/06/2021 12:33 AM <DIR> six-1.15.0.dist-info
05/05/2021 12:45 AM 34,159 six.py
05/06/2021 12:37 AM <DIR> urllib3
05/06/2021 12:33 AM <DIR> urllib3-1.25.9.dist-info
05/06/2021 12:37 AM <DIR> urllib3-1.26.4.dist-info
05/05/2021 12:45 AM 47,237 validate.py
05/06/2021 12:33 AM <DIR> win32
05/06/2021 12:34 AM <DIR> win32com
05/06/2021 12:33 AM <DIR> win32comext
05/06/2021 12:33 AM <DIR> zope
02/02/2021 07:10 PM 529 zope.component-4.6.1-py3.8-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.component-4.6.1.dist-info
05/05/2021 12:45 AM 529 zope.component-5.0.0-py3.8-nspkg.pth
05/06/2021 12:37 AM <DIR> zope.component-5.0.0.dist-info
02/02/2021 07:10 PM 529 zope.deferredimport-4.3.1-py3.7-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.deferredimport-4.3.1.dist-info
02/02/2021 07:10 PM 529 zope.deprecation-4.4.0-py3.7-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.deprecation-4.4.0.dist-info
02/02/2021 07:10 PM 529 zope.event-4.4-py2.7-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.event-4.4.dist-info
05/05/2021 12:45 AM 529 zope.event-4.5.0-py3.6-nspkg.pth
05/06/2021 12:37 AM <DIR> zope.event-4.5.0.dist-info
05/05/2021 12:45 AM 530 zope.hookable-5.0.1-py3.8-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.hookable-5.0.1.dist-info
02/02/2021 07:10 PM 530 zope.interface-5.1.0-py3.8-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.interface-5.1.0.dist-info
05/05/2021 12:45 AM 530 zope.interface-5.4.0-py3.8-nspkg.pth
05/06/2021 12:37 AM <DIR> zope.interface-5.4.0.dist-info
02/02/2021 07:10 PM 530 zope.proxy-4.3.5-py3.8-nspkg.pth
05/06/2021 12:33 AM <DIR> zope.proxy-4.3.5.dist-info
05/05/2021 12:45 AM 146,944 _cffi_backend.cp38-win32.pyd
05/06/2021 12:37 AM <DIR> _distutils_hack
05/05/2021 12:45 AM 21 _version.py
05/06/2021 12:37 AM <DIR> __pycache__
24 File(s) 3,081,221 bytes
77 Dir(s) 72,294,338,560 bytes free
</details>

**Comment by bmw** [comment]
@adferrand, you might be interested in this.

I wonder if instead of running the uninstaller as part of the auto-update script in https://github.com/certbot/certbot/pull/7539, we could modify the NSIS template to (at least effectively) uninstall a previous Certbot installation if it exists. I'm not sure how hard that is to do, but that approach could both fix this problem and work for the auto-update mechanism.

**Comment by adferrand** [comment]
I think it is quite doable to factor some of the cleanup logic in the NSIS template to be used both by the uninstall section and the initialization of the installation procedure indeed.

If you create a dedicated issue you can assign me or I can work directly with this bug report.
